### PR TITLE
feat(EvenTeams): add new config to improve the balance

### PR DESCRIPTION
**Pull request by Helias**

### CODE REVIEW

File: conf/CFBG.conf.dist [hunk 1]
```diff
@@ -27,8 +27,16 @@
 #       Description: Maximum players in party for enter queue
 #       Default: 3
 #
+#
+#    CFBG.InvitationType
+#        Description: Set Battleground invitation type.
+#        Default:     0 - (Normal, Invite as much players to battlegrounds as queued,
+#                          Don't bother with balance)
+#                     1 - (Experimental, Don't allow to invite much more players
+#                          of one faction)
```
Comment (FrancescoBorzi) [inline conf/CFBG.conf.dist]: you need to better explain this:

```
#        Default:     0 - (Normal, allow one team to have one more player than the other team)
#                     1 - (Experimental, only allow even teams)
```

File: src/CFBG.cpp [hunk 2]
```diff
@@ -612,6 +618,48 @@ bool CFBG::FillPlayersToCFBG(BattlegroundQueue* bgqueue, Battleground* bg, const
...
+    sLog->outError("BG COUNT %u", bg->GetPlayersSize());
+    sLog->outError("BG QUEUE %u", bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].size());
+    sLog->outError("BG balance %u", (bg->GetPlayersSize()+bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].size()) % 2 != 0);
+    sLog->outError("###########################################");
+
+    // do not allow to have more player in one faction
+    if (IsEnableInvitationType())
+    {
+        // if there is an even size of players in BG and only one in queue do not allow to join the BG
+        if (bg->GetPlayersSize() % 2 == 0 && bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].size() == 1) {
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp]: let's add one more conf, a `threshold`, such that:

```
bg->GetPlayersSize() < threshold && bg->GetPlayersSize() % 2 .....
```

File: src/CFBG.cpp [hunk 3]
```diff
@@ -612,6 +618,48 @@ bool CFBG::FillPlayersToCFBG(BattlegroundQueue* bgqueue, Battleground* bg, const
...
+            return false;
+        }
+
+
+        // if the sum of the players in BG and the players in queue is odd join one alliance and one horde per time, otherwise continue above and return true
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp]: `otherwise continue above and return true` ?? what does it mean?

File: src/CFBG_SC.cpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- *  Copyright (С) since 2019 Andrei Guluaev (Winfidonarleyan/Kargatum) https://github.com/Winfidonarleyan 
+ *  Copyright (С) since 2019 Andrei Guluaev (Winfidonarleyan/Kargatum) https://github.com/Winfidonarleyan
```
Comment (FrancescoBorzi) [inline src/CFBG_SC.cpp:2]: Let's also add the AC copyright as more people are working on it, but keep the original author who is @Winfidonarleyan

File: conf/CFBG.conf.dist [hunk 5]
```diff
@@ -27,8 +27,16 @@
 #       Description: Maximum players in party for enter queue
 #       Default: 3
 #
+#
+#    CFBG.InvitationType
```
Comment (FrancescoBorzi) [inline conf/CFBG.conf.dist]: let's rename it in something more meaningful... 

```
CFBG.EvenTeams
```

File: conf/CFBG.conf.dist [hunk 6]
```diff
@@ -27,8 +28,22 @@
 #       Description: Maximum players in party for enter queue
 #       Default: 3
 #
+#   CFBG.EvenTeams
```
Comment (FrancescoBorzi) [inline conf/CFBG.conf.dist]: rename to `CFBG.EvenTeams.Enabled`

File: conf/CFBG.conf.dist [hunk 7]
```diff
@@ -27,8 +28,22 @@
...
+#        Description: Set Battleground invitation type.
+#        Default:     0 - (Normal, Invite as much players to battlegrounds as queued,
+#                          Don't bother with balance)
+#                     1 - (Don't allow to invite much more players of one faction.
+#                          Keep an even quantity of players in Battleground)
```
Comment (FrancescoBorzi) [inline conf/CFBG.conf.dist]:
```
#        Default:     0 - (Disabled, one team can have one extra player if an odd number of players queues)
#                     1 - (Enabled. Get even teams.)
```

File: conf/CFBG.conf.dist [hunk 8]
```diff
@@ -27,8 +28,22 @@
...
+#
+#   CFBG.EvenTeams.MaxPlayersThreshold
+#       Description: team member quantity until the EvenTeams rule is valid
+#       Default: 5 - (Quantity of players after the EvenTeams rule will be disabled in the BG)
```
Comment (FrancescoBorzi) [inline conf/CFBG.conf.dist]: is this 5 per team or in total? It's not clear.

`Default: 5 - (Quantity of players per team after the EvenTeams rule will be ignored)`

File: src/CFBG.cpp [hunk 9]
```diff
@@ -42,6 +45,16 @@ bool CFBG::IsEnableBalancedTeams()
     return _IsEnableBalancedTeams;
 }
 
+bool CFBG::IsEnableEvenTeams()
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp:48]: `IsEvenTeamsEnabled`

File: src/CFBG.cpp [hunk 10]
```diff
@@ -42,6 +45,16 @@ bool CFBG::IsEnableBalancedTeams()
...
+{
+    return _IsEnableEvenTeams;
+}
+
+uint32 CFBG::MaxPlayersThreshold()
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp]: `EvenTeamsMaxPlayersThreshold`

File: src/CFBG.cpp [hunk 11]
```diff
@@ -607,10 +620,48 @@ bool CFBG::FillPlayersToCFBG(BattlegroundQueue* bgqueue, Battleground* bg, const
...
     bgqueue->m_SelectionPools[TEAM_HORDE].Init();
 
-    // quick check if nothing we can do:
-    if (!sBattlegroundMgr->isTesting())
-        if ((aliFree > hordeFree && bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].empty()))
+    uint32 bgPlayersSize = bg->GetPlayersSize();
+
+    // do not allow to have more player in one faction
+    // if treshold is enabled and if the current players quantity inside the BG is greater than the treshold
+    if (IsEnableEvenTeams() && !(MaxPlayersThreshold() > 0 && bgPlayersSize >= MaxPlayersThreshold()*2))
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp]: why `MaxPlayersThreshold() > 0` ?

File: src/CFBG.cpp [hunk 12]
```diff
@@ -607,10 +620,48 @@ bool CFBG::FillPlayersToCFBG(BattlegroundQueue* bgqueue, Battleground* bg, const
...
+    {
+        uint32 bgQueueSize = bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].size();
+
+        // if there is an even size of players in BG and only one in queue do not allow to join the BG
+        if (bgPlayersSize % 2 == 0 && bgQueueSize == 1) {
             return false;
+        }
+
+        // if the sum of the players in BG and the players in queue is odd, add all in BG except one
+        if ((bgPlayersSize+bgQueueSize) % 2 != 0) {
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp]: spaces around `+`

File: src/CFBG.cpp [hunk 13]
```diff
@@ -607,10 +620,48 @@ bool CFBG::FillPlayersToCFBG(BattlegroundQueue* bgqueue, Battleground* bg, const
...
+
+            return true;
+        }
+    }
+
+    // quick check if nothing we can do:
+    if (!sBattlegroundMgr->isTesting() && aliFree > hordeFree && bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].empty())
+    {
+        return false;
+    }
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp:666]: maybe add a comment like:

```cpp
// if CFBG.EvenTeams is disabled:
```

File: src/CFBG.cpp [hunk 14]
```diff
@@ -607,10 +620,48 @@ bool CFBG::FillPlayersToCFBG(BattlegroundQueue* bgqueue, Battleground* bg, const
     bgqueue->m_SelectionPools[TEAM_ALLIANCE].Init();
     bgqueue->m_SelectionPools[TEAM_HORDE].Init();
 
-    // quick check if nothing we can do:
-    if (!sBattlegroundMgr->isTesting())
-        if ((aliFree > hordeFree && bgqueue->m_QueuedGroups[bracket_id][BG_QUEUE_CFBG].empty()))
+    uint32 bgPlayersSize = bg->GetPlayersSize();
+
+    // do not allow to have more player in one faction
```
Comment (FrancescoBorzi) [inline src/CFBG.cpp]:
```cpp
// if CFBG.EvenTeams is enabled, do not allow to have more player in one faction:
```

File: src/CFBG.cpp (hunk 11 above)
Comment (Helias) [reply src/CFBG.cpp]: To check if it is enabled or not

Comment (FrancescoBorzi) [reply src/CFBG.cpp]: ok

File: src/CFBG.cpp (hunk 9 above)
Comment (Helias) [reply src/CFBG.cpp:48]: it's to be coherent with the others config variables

File: conf/CFBG.conf.dist (hunk 8 above)
Comment (Helias) [reply conf/CFBG.conf.dist]: per team

**Review by FrancescoBorzi — APPROVED** [review]

**Pull request merged**